opnode/node/node.go

```diff
@@ -0,0 +1,97 @@
+package node
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/rpc"
+	"time"
+)
+
+type OpNodeCmd struct {
+	L1NodeAddr   string `ask:"--l1" help:"Address of L1 User JSON-RPC endpoint to use (eth namespace required)"`
+	L2EngineAddr string `ask:"--l2" help:"Address of L2 Engine JSON-RPC endpoint to use (engine and eth namespace required)"`
+
+	LogCmd `ask:".log" help:"Log configuration"`
+
+	// TODO: multi-addrs option
+	// TODO: bootnodes option
+
+	log Logger
+
+	l1Node   *rpc.Client
+	l2Engine *rpc.Client
+
+	ctx   context.Context
+	close chan chan error
+}
+
+func (c *OpNodeCmd) Default() {
+	c.L1NodeAddr = "http://127.0.0.1:8545"
+	c.L2EngineAddr = "http://127.0.0.1:8551"
+}
+
+func (c *OpNodeCmd) Help() string {
+	return "Run optimism node"
+}
+
+func (c *OpNodeCmd) Run(ctx context.Context, args ...string) error {
+	log := c.LogCmd.Create()
+	c.log = log
+	c.ctx = ctx
+
+	// L1 exec engine: read-only, to update L2 consensus with
+	l1Node, err := rpc.DialContext(ctx, c.L1NodeAddr)
+	if err != nil {
+		return err
+	}
+	c.l1Node = l1Node
+
+	// L2 exec engine: updated by this OpNode (L2 consensus layer node)
+	engine, err := rpc.DialContext(ctx, c.L2EngineAddr)
+	if err != nil {
+		return err
+	}
+	c.l2Engine = engine
+
+	// TODO: maybe spin up an API server
+	//  (to get debug data, change runtime settings like logging, serve pprof, get peering info, node health, etc.)
+
+	c.close = make(chan chan error)
+
+	go c.RunNode()
+
+	return nil
+}
+
+// background process
+func (c *OpNodeCmd) RunNode() {
+	c.log.Info("started OpNode")
+
+	heartbeat := time.NewTicker(time.Millisecond * 700)
+	defer heartbeat.Stop()
+
+	for {
+		select {
+		case <-heartbeat.C:
+			// TODO poll data, process blocks, etc.
```

Comment:
This would also include reorging the L2 chain, is that right? 

Also, could consider adding a comment or a placeholder function to explicitly call out the handling of reorgs will be done in the driver. I just want to make sure that as we implement we **do not** put off handling reorgs till the end and instead are thinking about graceful reorg logic from the beginning.

Reply:
I'll implement collection of L1 data in a later PR to keep the size of this one reasonable. But yes, definitely want to handle reorgs. And subscribing, instead polling, would be neat too. I already put a TODO for it a few lines down. :+1: 

> // TODO: open a channel with L1 RPC to listen for new blocks and reorgs?